### ISSUE TITLE
EFS can't be mounted in 0.6.0

### ISSUE THREAD

**Issue by mhrabovcin** [codedellemc/rexray#609]
Trying to mount EFS volume by its name doesn't work in `0.6.0`. Mounting works with version built manually before this commit - https://github.com/emccode/rexray/commit/c68e7d706e5167af9ffa5423b82325b8f47ba7ea

**Comment by mhrabovcin** [comment]
Rexray version:

```
[root@ip-10-0-0-4 ~]# ./rexray version            
REX-Ray                                           
-------                                           
Binary: /root/rexray                              
SemVer: 0.6.0                                     
OsArch: Linux-x86_64                              
Branch: v0.6.0                                    
Commit: 429cfdf7272cebc5911b2865128ad37a6da5a203  
Formed: Fri, 21 Oct 2016 02:06:45 UTC             

libStorage                                        
----------                                        
SemVer: 0.3.1                                     
OsArch: Linux-x86_64                              
Branch: v0.6.0                                    
Commit: a3a561a1b94bb8b7efb4ae998540c457deb39692  
Formed: Fri, 21 Oct 2016 02:05:09 UTC
```

**Comment by kasisnu** [comment]
Some more debugging info:

```[root@ip-10-0-0-4 ~]# ./rexray version
REX-Ray
-------
Binary: /root/rexray
SemVer: 0.6.0
OsArch: Linux-x86_64
Branch: v0.6.0
Commit: 429cfdf7272cebc5911b2865128ad37a6da5a203
Formed: Fri, 21 Oct 2016 02:06:45 UTC

libStorage
----------
SemVer: 0.3.1
OsArch: Linux-x86_64
Branch: v0.6.0
Commit: a3a561a1b94bb8b7efb4ae998540c457deb39692
Formed: Fri, 21 Oct 2016 02:05:09 UTC
 
[root@ip-10-0-0-4 ~]# ./rexray volume ls
ID           Name       Status     Size
fs-53ad791a  wutututut  available  6144


[root@ip-10-0-0-4 ~]# ./rexray volume mount wutututut --service efs
FATA[0000] no volumes found

```

With http response logging enabled, we've confirmed volumes are being returned by the libstorage client

```e=objectivefs storageDriver=libstorage time=1477475869228 txCR=1477475869 txID=80697ee2-949c-4749-7d12-85b797132905
DEBU[0000] xli localdevices success                      host=tcp://127.0.0.1:7979 instanceID=efs=subnet-1a1b9753 integrationDriver=docker osDriver=linux server=zest-braid-na servic
e=objectivefs storageDriver=libstorage time=1477475869228 txCR=1477475869 txID=80697ee2-949c-4749-7d12-85b797132905
INFO[0000]
INFO[0000]     -------------------------- HTTP RESPONSE (CLIENT) -------------------------
INFO[0000]     HTTP/1.1 200 OK
INFO[0000]     Content-Length: 109
INFO[0000]     Content-Type: application/json
INFO[0000]     Date: Wed, 26 Oct 2016 09:57:49 GMT
INFO[0000]     Libstorage-Servername: zest-braid-na
INFO[0000]
INFO[0000]     {
INFO[0000]       "fs-53ad791a": {
INFO[0000]         "name": "wutututut",
INFO[0000]         "size": 6144,
INFO[0000]         "id": "fs-53ad791a",
INFO[0000]         "type": ""
INFO[0000]       }
INFO[0000]     }
FATA[0000] no volumes found```

**Comment by akutz** [comment]
Hi @mhrabovcin and @kasisnu,

I am looking at this today. Thanks!

**Comment by akutz** [comment]
Hi @mhrabovcin / @kasisnu  (cc @cduchesne),

Okay, so the issue is a "protection" I added to the RR mount command. It asserts a volume should be attached prior to being mounted. Obviously I need to reexamine that assertion for NFS providers.

However, I'm running into an additional issue. I'm now attempting to attach the EFS volume and getting an error about [an invalid security ID](https://gist.github.com/akutz/2877a8e21b6d1b7c64a00f35dbcd2f44).

According to the EFS driver [docs](http://libstorage.readthedocs.io/en/stable/user-guide/storage-providers/#aws-efs) if no security groups are specified the default VPC should be used. So shouldn't things work without an explicit security group list?

**Comment by akutz** [comment]
Okay guys, I have it working. It will require an update to REX-Ray if I want to continue to use a volume's attached status to determine whether or not it can be mounted. Currently the volume query to match volume names/IDs for mutative ops does not request attachment data. For EBS it was always sent regardless, so I didn't encounter the bug. I updated the RR code on an EC2 instance and [it works](https://gist.github.com/akutz/edaf5724595b4293c062f8b4c7e5b17e) with EFS just fine now.

### Showing as Attached
```bash
[root@ip-172-31-25-144 rexray]$ rexray -l error volume ls --service efs
ID           Name    Status    Size
fs-03ec16aa  builds  attached  6144
```

### Mounting
```bash
[root@ip-172-31-25-144 rexray]$ rexray -l error volume mount builds --service efs
ID           Name    Status    Size  Path
fs-03ec16aa  builds  attached  6144  /var/lib/rexray/volumes/builds/data
```

**Comment by akutz** [comment]
So what does everyone think -- should we continue to determine if a volume can be mounted/unmounted/attached/detached/etc. based on whether it is attached already or not? And if so, should we just go ahead and always return attachment data par for the course? Meaning the RR CLI will always send the `attachments=true` flag?

**Comment by mhrabovcin** [comment]
Since `attachments` flag is optional I think it makes sense for RR to specifically ask for data that it requires in order to successfully complete the operation. Checking for attachment status before mounting sound reasonable.

**Comment by akutz** [comment]
Hi @mhrabovcin & @kasisnu,

I need your help. The libStorage [release branch for 0.3.3](https://github.com/codedellemc/libstorage/tree/release/0.3.3) in concert with the REX-Ray [0.6.1 release branch](https://github.com/codedellemc/rexray/tree/release/0.6.1) solves the mounting issue. However, I've run into a few issues:

- The Attach op does not work unless I manually specify a security group in the config file. Did default security groups ever work?
- Why was a CSV used for security groups instead of a YAML array. Just curious.
- The Detach op doesn't work. It doesn't fail, but it doesn't work. Nothing happens.
- The Unmount op fails, but there are no errors on the client or server side. Nothing. The only error is the one saying the op failed, but nothing contained in or before that gives hint as to why. Did the unmount op ever work?

My intent on the "did XYZ ever work" is not to offend or place blame. I'm just not a user of the EFS driver, and this work to patch the issue of not mounting in RR 0.6.0 was my first real foray into the platform. Thus I do not know if things that are broken are because of changes I or others have made or these are existing issues.

Thank you in advance for your help!

**Comment by akutz** [comment]
By the way, please find the new `examples` [package](https://github.com/codedellemc/libstorage/tree/release/0.3.3/examples/ec2/vagrant) for assistance in bringing an environment for testing EBS/EFS online using Vagrant.

**Comment by kasisnu** [comment]
Hey,

Sorry you're having trouble with this.

Talked with Martin about this.

> The Attach op does not work unless I manually specify a security group in the config file. Did default security groups ever work?

The security group in the config file _is_ the default security group that is used when creating mount targets. We should look at failing the storage driver during `Init()` if that isn't specified. Martin pointed out that the documentation is misleading in that we don't lookup the default vpc security group. This should be updated.

> Why was a CSV used for security groups instead of a YAML array. Just curious.

This can(should) be changed.

> The Detach op doesn't work. It doesn't fail, but it doesn't work. Nothing happens.

The Detach op isn't supposed to do anything - or that it can't in case of a shared filesystem.
Because the detach call is called in the context of a host, there can be other hosts using the same mount target so it is not safe to remove mount targets in a detach call. We remove them in a delete call because in that case the intent is clear - to remove the filesystem entirely. It is also not an error condition. It truly is a no-op. Maybe we can better express that in the code?

> The Unmount op fails, but there are no errors on the client or server side. Nothing. The only error is the one saying the op failed, but nothing contained in or before that gives hint as to why. Did the unmount op ever work?

This is the most interesting bit to me. This can happen if the volume was removed out-of-band, ex. the mount target was deleted and the nfs client is then essentially stuck "waiting", or if some process is using files on the volume. This seems like something that the storage driver isn't involved in though.

We're looking at making changes for #1 & #2.

Could you list the steps to get an unmount to fail or is this all unmounts?

Thanks

**Comment by akutz** [comment]
> Maybe we can better express that in the code?

You expressed the Detach op being a no/op operation fine. I just reviewed the wrong function when looking at it for some reason. That's on me.

I will get some steps together to repro the unmount scenario. Thanks!

**Comment by akutz** [comment]
Hi @mhrabovcin / @kasisnu,

I just opened libStorage PR [\#343](https://github.com/codedellemc/libstorage/pull/343) that touches on some of this. At least starts to touch upon it anyway. I'd love to get your thoughts on it.

**Comment by akutz** [comment]
Closed with #635.
